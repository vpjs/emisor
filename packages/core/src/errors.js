export class EmisorTypeError extends TypeError {
  constructor(param, expect, given) {
    super(`${param} has to be a ${expect}, ${typeof given} given`);
  }
}

export class EmisorError extends Error {}

/**
 * Error object
 */
export class EmisorPluginError extends EmisorError {}


export class EmisorPluginTypeError extends EmisorTypeError {}