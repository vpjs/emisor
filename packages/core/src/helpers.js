function getName (obj) {
  return ({}).toString.call(obj);
}

/**
 * is given value a symbol
 * @param {*} value
 * @returns {boolean}  
 */
export function isSymbol (value) {
  return typeof value === 'symbol' || getName(value) == '[object Symbol]';
}

/**
 * is given value a RegExp
 * @param {*} value
 * @returns {boolean}  
 */
export function isRegExp (value) {
  return getName(value) == '[object RegExp]';
}
/**
 * check of given value is a string
 * @param {*} value
 * @returns {boolean} 
 */
export function isString (value) {
  return typeof value === 'string' || getName(value) == '[object String]';
}

/**
 * check of given value is a function
 * @param {*} value
 * @returns {boolean} 
 */
export function isFunction (value) {
  return typeof value === 'function';
}