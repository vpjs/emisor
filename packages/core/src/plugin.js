/**
 * @typedef {import('./hook').EmisorHook} EmisorHook
 * @typedef {import('./hook').EmisorHookAll} EmisorHookAll
 * @typedef {import('./hook').EmisorHookEventStr} EmisorHookEventStr
 */

import { EmisorPluginError } from './errors';

/**
 * @typedef {object} EmisorPluginHook 
 * @prop {EmisorHook["pluginApi"]} beforeOn
 * @prop {EmisorHook["pluginApi"]} afterOn
 * @prop {EmisorHook["pluginApi"]} beforePublish
 * @prop {EmisorHook["pluginApi"]} afterPublish
 * @prop {EmisorHook["pluginApi"]} beforeOff
 * @prop {EmisorHook["pluginApi"]} afterOff
 * @prop {EmisorHookAll["pluginApi"]} onEmit
 * @prop {EmisorHookEventStr["pluginApi"]} eventStr
 */

/**
 * @typedef {object} IEmisorPlugin
 * @prop {(EmisorPluginHook)=>void} install - install plugin
 */

const OVERWRITE_PAYLOAD_KEY = Symbol();
const OVERWRITE_HANDLER_KEY = Symbol();
const BREAK_KEY = Symbol();
const KILL_KEY = Symbol();
const REMOVE_TAG = Symbol();
const ADD_TAG = Symbol();
/**
 * @implements {IEmisorPlugin}
 */
export class EmisorPlugin {

  /**
    * Result key that should contain the new payload
    * Can only be used by `beforePublish`and `onEmit`hooks,
    * and `afterPublish` but will only affect payload for other hooks
    * @example (payload, $event) => ({[EmisorPlugin.OVERWRITE_PAYLOAD_KEY]: 'new payload'})
    */
  static get OVERWRITE_PAYLOAD_KEY () {
    return OVERWRITE_PAYLOAD_KEY;
  }

  /**
   * Result key that should contain the new event handler
   * Can only be used by `beforeOn`
   * @example (payload, $event) => ({[EmisorPlugin.OVERWRITE_HANDLER_KEY]: () => {}})
   */
  static get OVERWRITE_HANDLER_KEY () {
    return OVERWRITE_HANDLER_KEY;
  }

  /**
   * This will break execution of all other hooks
   */
  static get BREAK_KEY () {
    return BREAK_KEY;
  }

  /**
   * This will break execution of all other hooks and stop further execution emit
   */
  static get KILL_KEY () {
    return KILL_KEY;
  }

  static get REMOVE_TAG () {
    return REMOVE_TAG;
  }

  static get ADD_TAG () {
    return ADD_TAG;
  }

  install () {
    throw new EmisorPluginError('Plugin is missing a install method');
  }
}