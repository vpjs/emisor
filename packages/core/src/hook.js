import { EmisorPluginTypeError, EmisorPluginError } from './errors';
import { isSymbol, isString, isRegExp, isFunction } from './helpers';

/** @typedef {import('./').EmisorEvent} EmisorEvent */
/** @typedef {import('./').EmisorEventObject} EmisorEventObject */

/**
 * @typedef {object} EmisorHookMeta
 * @prop {EmisorEventObject} meta.event
 * @prop {*} [meta.payload]
 * @prop {object} meta.storage
 * @prop {*} [meta.options]
 */

/**
 * @callback EmisorHookCallback
 * @param {EmisorHookMeta} meta
 * @param {import('./').EmisorHookAPI} Emisor
 * @returns {object}
 */

/**
 * @callback EmisorHookFnc
 * @param {EmisorEventObject} event
 * @param {*} [payload]
 * @return {EmisorHookCallback}
 */

export class EmisorHookAll {
  /**
   * @type {Set<EmisorHookCallback>}
   */
  #allCallbacks = new Set();

  get pluginApi () {
    return {
      /**
       * Register a hook all events
       * @param {EmisorHookCallback} callback 
       */
      all: (callback) => {
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#allCallbacks.add(callback);
      } 
    };
  }

  /**
   * 
   * @param {import('./').EmisorHookAPI} Emisor
   * @returns {EmisorHookFnc[]}
   */
  getHooks (Emisor) {
    return Array.from(this.#allCallbacks)
    .map((callback) => {
      let  storage = {};
      return (event, payload) => callback({
        event,
        payload,
        storage
      }, Emisor);
    });
  }
}



export class EmisorHook  {
  #index = 0
  /**
   * @type {Map<string|Symbol|{all: true}, EmisorHookCallback>}
   */
  #keyCallbacks = new Map();

  /**
   * @type {Set<EmisorHookCallback>}
   */
  #allCallbacks = new Set();

  get pluginApi () {
    return {
      /**
       * Register a hook based on a key
       * @param {string|Symbol} key
       * @param {EmisorHookCallback} callback 
       */
      key: (key, callback) => {
        if (!isString(key) && !isSymbol(key)) {
          throw new EmisorPluginTypeError('key', 'string or Symbol', key);
        }
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#keyCallbacks.set(key, {
          callback,
          index: this.#index++
        });
      },
      /**
       * Register a hook all events
       * @param {EmisorHookCallback} callback 
       */
      all: (callback) => {
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#allCallbacks.add({
          callback,
          index: this.#index++
        });
      }
    };
  }

  /**
   * @param {object} options 
   * @param {import('./').EmisorHookAPI} Emisor
   * @returns {EmisorHookFnc[]}
   */
  getHooks (options, Emisor) {
    return Object.entries(options)
    .filter(([key]) => this.#keyCallbacks.has(key))
    .map(([key, $options]) => ({
      ...this.#keyCallbacks.get(key),
      $options
    }))
    //add "all" callbacks
    .concat([...this.#allCallbacks])
    //sort
    .sort(({index: a}, {index: b}) => a -  b)
    ///create hooks
    .map(({callback, $options}) => {
      let storage = {};
      /**
         * @param {import('./').EmisorEvent} event
         * @param {*} [payload]
         * @returns {EmisorHookCallBack}
         */
      return (event, payload) => callback({
        storage,
        payload,
        event,
        ...($options ? {options:  $options} : null)
      }, Emisor);

    });
  }
}

const PREFIX_ALLOWED = '<>[]{}()/\\:;.,\'"-=+~*^!@#$%&?_`';

/**
 * @callback EmisorHookEventStrCallBack
 * @param {string} $event
 * @return {Object<string|Symbol, any>}
 */
export class EmisorHookEventStr {
  /**
   * @type {string}
   */
  #postfixSeparator

  /**
   * @type {RegExp}
   */
  #allowedPrefixRegex

  /**
   * @type {string}
   */
  #postfixDivider

  /**
   * @type {Map<RegExp, EmisorHookEventStrCallBack>}
   */
  #postfixCallbacks = new Map();

  /**
   * @type {Map<RegExp, EmisorHookEventStrCallBack>}
   */
  #prefixCallbacks = new Map();

  /**
   * @param {string} s
   */
  set postfixSeparator(s) {
    if (!isString(s)) {
      throw new EmisorPluginTypeError('postfixSeparator', 'string', s);
    }
    if(s.length !== 1) {
      throw new EmisorPluginError('postfixSeparator can not be longer then 1');
    }
    
    this.#postfixSeparator = s;
    this.#updateAllowedPrefixRegex();
    
  }

  /**
   * @param {string} s
   */
  set postfixDivider(s) {
    if (!isString(s)) {
      throw new EmisorPluginTypeError('postfixDivider', 'string', s);
    }
    if(s.length !== 1) {
      throw new EmisorPluginError('postfixDivider can not be longer then 1');
    }
    this.#postfixDivider = s;
    this.#updateAllowedPrefixRegex();
  }

  get pluginApi () {
    return {
      /**
       * Register a postfix hook
       * @param {RegExp} regex
       * @param {EmisorHookEventStrCallBack} callback
       */
      postfix: (regex, callback) => {
        if (!isRegExp(regex)) {
          throw new EmisorPluginTypeError('regex', 'RegExp', regex);
        }
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        this.#postfixCallbacks.set(regex, callback);
      },

      /**
       * Register a prefix hook
       * @param {string} char any char that is not a-z, 0-9 or postfix separator
       * @param {EmisorHookEventStrCallBack} callback
       */
      prefix: (char, callback) => {
        if (!isString(char)) {
          throw new EmisorPluginTypeError('char', 'string', char);
        }
        if (char.length !== 1) {
          throw new EmisorPluginError('char can not be longer then 1');
        }
        if (!this.#allowedPrefixRegex.test(char)) {
          throw new EmisorPluginError(`char "${char}" is not allowed`);
        }
        if (!isFunction(callback)) {
          throw new EmisorPluginTypeError('callback', 'function', callback);
        }
        if (this.#prefixCallbacks.has(char)) {
          throw new EmisorPluginError(`There is already a hook registered with "${char}"`);
        }
        this.#prefixCallbacks.set(char, callback);
      }
    };
  }

  constructor () {
    this.#updateAllowedPrefixRegex();
  }

  /**
   * Parse event string
   * @param {string} event 
   * @return {{event:string, options: Object<string|Symbol, any>}}
   */
  parseStr (event) {
    let postfix = this.#postfixStr(event),
        prefix = this.#prefixStr(postfix.event);
    return {
      event: prefix.event,
      options: {
        ...postfix.options,
        ...prefix.options
      }
    };
  }

  /**
   * Parse prefix string and runs prefix hooks if any
   * @param {string} rawEvent 
   * @return {{event:string, options: Object<string|Symbol, any>}}
   */
  #prefixStr (rawEvent) {
    let char = rawEvent[0],
        hook = this.#prefixCallbacks.get(char);
    //run hook
    if (hook) {
      return {
        event: rawEvent.substr(1),
        options: {
          ...hook(char)
        }
      };
    }
    return {
      event: rawEvent,
      options: {}
    };
  }

  /**
   * Parse postfix string and runs postfix hooks if any
   * @param {string} rawEvent 
   * @return {{event:string, options: Object<string|Symbol, any>}}
   */
  #postfixStr (rawEvent) {
    let options = {},
        [event, ...postfix] = rawEvent.split(this.#postfixDivider);
    postfix = postfix.join(this.#postfixDivider);
    //only when event has a postfix
    if (postfix) {
      postfix.split(this.#postfixSeparator)
      .forEach((plugin) => {
        this.#postfixCallbacks.forEach((hook, reg) => {
          if (reg.test(plugin)) {
            options = {
              ...options,
              ...hook(plugin)
            };
          }
        });
      });
      
    }
    return {
      event,
      options
    };
  }

  /**
   * update allowed prefix regex
   */
  #updateAllowedPrefixRegex () {
    let needsEscaping = [
          '-',
          ']'
        ],
        disallowed = [
          this.#postfixDivider,
          this.#postfixSeparator,
          '*' //wild card
        ],
        allowed = PREFIX_ALLOWED.split('')
        .filter((char) => !disallowed.includes(char))
        .map((char) => {
          if (needsEscaping.includes(char)) {
            return `\\${char}`;
          }
          return char;
        })
        .join('');
    this.#allowedPrefixRegex = new RegExp(`[${allowed}]`);
  }
}