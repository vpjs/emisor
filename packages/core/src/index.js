import { EmisorError, EmisorTypeError } from './errors';
import { isString, isSymbol, isFunction } from './helpers';
import { EmisorHook, EmisorHookAll, EmisorHookEventStr } from './hook';
import { EmisorPlugin } from './plugin';
export * as helpers from './helpers';

/**
 * @typedef {import('./hook').EmisorHookFnc} EmisorHookFnc
 */
/**
 * @typedef {import('./hook').EmisorHookMeta} EmisorHookMeta
 */
/**
 * @typedef {import('./plugin').IEmisorPlugin} IEmisorPlugin
 */

/**
 * @typedef {object} EmisorOptions
 * @prop {string} [nsSeparator='.'] namespace separator options
 * @prop {string} [postfixSeparator=':'] postfix separator
 * @prop {IEmisorPlugin[]} [plugins] plugins
 */
 
/**
 * @typedef {string|Symbol} EmisorEvent
 */

/**
  * @typedef {Object<string|number, any>} EmisorEventOptions
  */

 
/**
 * @typedef {Map<string, EmisorHookFnc[]>} EmisorPluginHooksMap
 */

/**
 * @typedef {Map<EmisorEventHandler, object>} EmisorSubsMap
 */

/**
 * @typedef {object} EmisorEventObject
 * @prop {string} [id] of event `1-1`
 * @prop {number} time unix timestamp of publish time
 * @prop {EmisorEvent} event that triggered the handler
 * @prop {EmisorEventHandler} [handler]
 * @prop {*} [tag]
 */


/**
  * @typedef {object} EmisorChainingAPI
  * @prop {EmisorCore["on"]} on
  * @prop {EmisorCore["emit"]} emit
  * @prop {EmisorCore["off"]} off
  */

/**
  * @typedef {object} EmisorHookAPI
  * @prop {EmisorCore["on"]} on
  * @prop {EmisorCore["emit"]} emit
  * @prop {EmisorCore["off"]} off
  * @prop {Function} rawEmit
  * @prop {Function} parseEvent
  */

/**
 * @callback EmisorEventHandler
 * @param {*} data
 * @param {EmisorEventObject} $event
 */

const DEFAULT_NS_SEPARATOR = '.';
const DEFAULT_POSTFIX_SEPARATOR = ',';
const DEFAULT_WILDCARD = '*';
const DEFAULT_POSTFIX_DIVIDER= '?';

let DEFAULT_ID_COUNTER = 0;


export const UNCAUGHT_EXCEPTIONS_EVENT = Symbol();

export default class EmisorCore {
  /**
   * @type {Map<EmisorEvent, EmisorSubsMap>}
   */
  #subs = new Map()

  /**
   * @type {string} namespace separator
   */
  #nsSeparator
  /**
   * Hook that will be trigger before subscribed to a event 
   */
  #beforeOnHook = new EmisorHook()
  /**
   * Hook that will be trigger after subscribed to a event 
   */
  #afterOnHook = new EmisorHook()
  /**
   * Hook that will be trigger before event publish
   */
  #beforePublishHook = new EmisorHook()
  /**
   * Hook that will be trigger after event publish
   */
  #afterPublishHook = new EmisorHook()

  /**
   * Hook that will be trigger after emit method is invoked
   */
  #onEmitHook = new EmisorHookAll()

  /**
   * Hook that will be trigger before event is unsubscribed
   */
  #beforeOffHook = new EmisorHook()
  /**
   * Hook that will be trigger after event is unsubscribed
   */
  #afterOffHook = new EmisorHook()

  #eventStrHook = new EmisorHookEventStr()

  /**
   * @type {EmisorHookAPI}
   */
  get #hookApi () {
    return {
      on: (...args) => this.on(...args),
      emit: (...args) => this.emit(...args),
      off: (...args) => this.off(...args),
      rawEmit: (...args) => this.#rawEmit(...args),
      parseEvent: (event) => this.#parseEvent(event)
    };
  }

  /**
   * 
   * @param {EmisorOptions} options 
   */
  constructor ({
    nsSeparator = DEFAULT_NS_SEPARATOR,
    postfixSeparator = DEFAULT_POSTFIX_SEPARATOR,
    plugins = []
  } = {}) {
    [
      ['nsSeparator', nsSeparator],
      ['postfixSeparator', postfixSeparator]
    ].forEach(([key, value]) => {
      if (!isString(value)) {
        throw new EmisorTypeError(key, 'string', value);
      }
      if (value.length !== 1) {
        throw new EmisorTypeError(`${key} has to have exactly a length of 1`);
      }
      if ([DEFAULT_WILDCARD, DEFAULT_POSTFIX_DIVIDER].includes(value)) {
        throw new EmisorError(`${key} can not be * or ?`);
      }
    });
    if (postfixSeparator === nsSeparator) {
      throw new EmisorError('nsSeparator and postfixSeparator can not be the same');
    }
    this.#nsSeparator = nsSeparator;
    this.#eventStrHook.postfixSeparator = postfixSeparator;
    this.#eventStrHook.postfixDivider = DEFAULT_POSTFIX_DIVIDER;

    plugins.forEach((plugin) => {
      plugin.install({
        beforeOn: this.#beforeOnHook.pluginApi,
        afterOn: this.#afterOnHook.pluginApi,
        beforePublish: this.#beforePublishHook.pluginApi,
        afterPublish: this.#afterPublishHook.pluginApi,
        beforeOff: this.#beforeOffHook.pluginApi,
        afterOff: this.#afterOffHook.pluginApi,
        onEmit: this.#onEmitHook.pluginApi,
        eventStr: this.#eventStrHook.pluginApi
      }, this.#chain());
    });
  }

  /**
   * Subscribe to a event
   * @param {EmisorEvent} event 
   * @param {EmisorEventHandler} handler 
   * @param {object} [options]
   * @returns {EmisorChainingAPI};
   */
  on (event, handler, options = {}) {
      

    if(isString(event)) {
      let result = this.#eventStrHook.parseStr(event);
      event = result.event;
      options = {
        ...result.options,
        ...options
      };
    }

    if(!this.#subs.has(event)) {
      this.#subs.set(event, new Map());
    }
    let $event = {
      event,
      handler
    };
    
    //before on hooks
    for (let hook of this.#beforeOnHook.getHooks(options, this.#hookApi)) {
      let result = hook($event);
      if (result?.[EmisorPlugin.OVERWRITE_HANDLER_KEY]) {
        handler = result[EmisorPlugin.OVERWRITE_HANDLER_KEY];
      }

      if (result?.[EmisorPlugin.BREAK_KEY]) {
        break;
      }
    }
    
    //subscribe
    this.#subs.get(event).set(handler, options);
    //after on hooks
    for (let hook of this.#afterOnHook.getHooks(options, this.#hookApi)) {
      let result = hook($event);
      if (result?.[EmisorPlugin.BREAK_KEY]) {
        break;
      }
    }
    return this.#chain();
  }
  /**
   * Unsubscribe
   * If no `event` is given, all subscribe will be unsubscribed 
   * if no `handler` is given all subscribe of the given `event` will be unsubscribed
   * @param {EmisorEvent} [event] unsubscribe to a specific event
   * @param {EmisorEventHandler} [handler] unsubscribe to a specific handler
   * @returns {EmisorChainingAPI}
   */
  off (event, handler) {
    //remove all
    if (!event) {
      this.#subs = new Map();
      return this.#chain();
    }
    let handlers = this.#subs.get(event);
    //if a event doesn't have any handlers we can don't unsubscribe anything 
    if(!handlers) {
      return this.#chain();
    }
    //remove all handlers of event
    if (!handler) {
      handlers.clear();
    } else {
      handlers.delete(handler);
    }
    return this.#chain();
  }

  /**
   * 
   * @param {EmisorEvent} event 
   * @param {*} [payload] 
   * @return {EmisorChainingAPI}
   */
  emit (event, payload) {
    return this.#chain (
      this.#runOnEmitHook(event, payload)
    );
  }

  async #runOnEmitHook (event, payload) {
    let {
      $event,
      $payload
    } = await this.#runHooks({ event}, payload, this.#onEmitHook.getHooks(this.#hookApi));
    //no event given or hook has killed execution
    if (!$event) {
      return;
    }
    this.#rawEmit($event.event, $payload);
  }

  /**
   * 
   * @param {EmisorEvent} event 
   * @param {*} [payload]
   * @param {EmisorEventHandler|EmisorEventHandler[]} [only]
   * @param {*} [tag]
   */
  async #rawEmit (event, payload, only, tags) {
    /** @type {Array<[EmisorEventHandler, EmisorPluginHooksMap]>} */
    let subs = [], 
        parsedEvents = this.#parseEvent(event),
        $event = {
          tags,
          event
        };
    
    // {generator: filterHooks, updateEvent} = this.#hooksGeneratorInit({}, payload, [
    //   () => only ? /** @type {EmisorEventHandler[]} **/ (only).includes(handler) : true
    // ]);
    //before emit
    only = /** @type {EmisorEventHandler[]} **/ (isFunction(only) ? [only] : only);
    parsedEvents.forEach((event) => {
      if(this.#subs.has(event)) {
        subs = [
          ...subs,
          ...Array.from(this.#subs.get(event))
        ];
      }
    });
    // let handlerIndex = 0
    // updateEvent({
    //   ...$event,
    //   handler: subs[++handlerIndex]
    // })
    // for await (let filter of filterHooks) {
    //   for 
    // }


    subs
    .filter(([handler]) => only ? /** @type {EmisorEventHandler[]} **/ (only).includes(handler) : true)
    .forEach(([handler, options]) => this.#publish({
      event: {
        ...$event,
        handler,
        time: Date.now(),
        id: `${DEFAULT_ID_COUNTER++}`
      },
      payload,
      options
    }));
  }

  /**
   * @param {{event: EmisorEvent, handler: EmisorEventHandler, payload: any, tag: any, options: any}} publishObj
   */
  async #publish ({event, payload, options}) {
    //before emit hooks
    let {
          $event, 
          $payload
        } = await this.#runHooks(event, payload, this.#beforePublishHook.getHooks(options, this.#hookApi)),
        { handler } = $event;

    //no event given or hook has killed execution
    if (!$event) {
      return;
    }

    //publish
    try {
      await handler($payload, $event);
    } catch (error) {
      this.#rawEmit(UNCAUGHT_EXCEPTIONS_EVENT, {
        error,
        event: $event,
        payload: $payload
      });
    } 
    //after emit hooks
    await this.#runHooks($event, payload,  this.#afterPublishHook.getHooks(options,this.#hookApi));
  }

  async #runHooks ($event, $payload, hooks) {
    for (let hook of hooks) {
      let result = await hook($event, $payload);
      //update payload
      if (result?.[EmisorPlugin.OVERWRITE_PAYLOAD_KEY]) {
        $payload = result[EmisorPlugin.OVERWRITE_PAYLOAD_KEY];
      }
      if (result?.[EmisorPlugin.KILL_KEY]) {
        return {};
      }
      //break
      if (result?.[EmisorPlugin.BREAK_KEY]) {
        break;
      }
    }
    return {$payload, $event};
  }

  /**
   * Emisor chain, that is also promise aware
   * @param {Promise} [promise]
   * @returns {EmisorChainingAPI}
   */
  #chain (promise) {
    let methods = ['on', 'emit', 'off'],
        wrappers ;
    if (!promise) {
      wrappers = methods.map((key) => [
        key,
        (...args) => this[key](...args)
      ]);
    } else {
      wrappers = methods.map((key) => [
        key,
        (...args) => {
          promise.finally(() => this[key](...args));
          return this.#chain(promise);
        }
      ]);
    }
    return Object.fromEntries(wrappers);
  }

  /**
   * @param {EmisorEvent} event
   * @returns {EmisorEvent[]} 
   */
  #parseEvent (event) {
    if (isSymbol(event)) {
      return [event, '*'];
    } else if (isString(event)) { 
      let ns = /** @type {string} **/ (event).split(this.#nsSeparator);
      return [
        event,
        `${event}${this.#nsSeparator}${DEFAULT_WILDCARD}`,
        ...ns.map((_, i, arr) => {
          let x = [...arr];
          x[i] = DEFAULT_WILDCARD;
          return [
            x.join(this.#nsSeparator),
            [...arr.slice(0,i), DEFAULT_WILDCARD].join(this.#nsSeparator)
          ];
        }).flat()
      ].filter((v, i, a) => a.indexOf(v) === i);
    }
  }

}

export {
  EmisorCore,
  EmisorPlugin
};