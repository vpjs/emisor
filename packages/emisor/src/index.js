import { EmisorCore } from '@emisor/core';
import { EmisorPluginCount } from '@emisor/plugin-count';
import { EmisorPluginHistory, MODE_DEFAULT_ALLOW, MODE_DEFAULT_DENY } from '@emisor/plugin-history';

/** @typedef {import('@emisor/core').EmisorEvent} EmisorEvent */
/** @typedef {import('@emisor/core').EmisorEventHandler} EmisorEventHandler */


//history modes
export const HISTORY_MODE_DEFAULT_ALLOW = MODE_DEFAULT_ALLOW;
export const HISTORY_MODE_DEFAULT_DENY = MODE_DEFAULT_DENY;

export class Emisor extends EmisorCore {
  #historyKey;
  #countKey;
  
  /**
   * @param {import('@emisor/core').EmisorOptions} options
   * @param {object} [pluginConfig]
   */
  constructor ({
    plugins = [], 
    ...options
  } = {}, {
    history = {}, 
    count = {}
  } = {}) {
    super({
      plugins: [
        new EmisorPluginCount(count),
        new EmisorPluginHistory(history),
        ...plugins
      ],
      ...options,
      chain: [
        ...(options?.addToChain || []),
        'once',
        'many',
        'history',
        'historyOnce'
      ],
    });
    this.#historyKey = history.key || 'history';
    this.#countKey = count.key || 'count';
  }

  /**
   * @param {EmisorEvent} event
   * @param {EmisorEventHandler} handler
   */
  once (event, handler) {
    return this.on(event, handler, {
      [this.#countKey]: 1
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {number} count
   * @param {EmisorEventHandler} handler
   */
  many (event, count, handler) {
    return this.on(event, handler, {
      [this.#countKey]: count
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {number|boolean|EmisorEventHandler} length
   * @param {EmisorEventHandler} handler
   */
  history (event, length, handler) {
    if (typeof length === 'function') {
      handler = length;
      length = true;
    }
    return this.on(event, handler, {
      [this.#historyKey]: length
    });
  }

  /**
   * @param {EmisorEvent} event
   * @param {EmisorEventHandler} handler
   */
  historyOnce (event, handler) {
    return this.on(event, handler, {
      [this.#historyKey]: true ,
      [this.#countKey]: 1
    });
  }
}