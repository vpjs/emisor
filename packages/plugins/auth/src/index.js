import { EmisorPlugin, helpers } from '@emisor/core';

/**
 * 
 */
export class EmisorPluginAuth extends EmisorPlugin {


  /**
   */
  constructor () {
    super();
  }

  /**
  * @param {import('@emisor/core').EmisorPluginHooks} hook
  */
  install ({}) {
    
  }
}