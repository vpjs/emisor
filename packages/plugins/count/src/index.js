//@ts-check
import { EmisorPlugin } from '@emisor/core';

const COUNT = Symbol('count');

export class EmisorPluginCount extends EmisorPlugin {
  
  /**
   * @param {{key: string}} [options]
   */
  constructor ({key = 'count'}) {
    super({key});
  }
  
  /**
   * @param {import('@emisor/core').EmisorCore} Emisor 
   * @param {import('@emisor/core').EmisorEventObject} $event 
   * @param {object} $config
   */
  afterEmit (Emisor, $config, _, $event) {
    //key should contain a number what is higher then 0
    if ($config[this.key] > 0) {
      //get account
      $config[COUNT] = $config[COUNT] ?? $config[this.key];
      if ($config[COUNT] === 1) {
        Emisor.off($event.event, $event.handler);
      } else {
        $config[COUNT]--;
      }
    }
  }

}