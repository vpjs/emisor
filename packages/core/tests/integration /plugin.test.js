import { EmisorCore, EmisorPlugin } from '../../src';

describe('Plugin hook interface', () => {

  test('plugin install hook signature', () => {
    let install = jest.fn();
    class Plugin extends EmisorPlugin {
      install = install
    }

    new EmisorCore({
      plugins: [
        new Plugin
      ]
    });

    expect(install).toMatchSnapshot();

  });
});

describe('eventStr hooks', () => {
  test('eventStr.postfix', () => {  
    let regexp = /<(?<test>\d+)>/, 
        postfixCallback = jest.fn(($event) => {
          let {test} = regexp.exec($event).groups;
          return {test};
        }),
        keyCallback = jest.fn();
    class Plugin extends EmisorPlugin {
      install({eventStr, beforeOn}) {
        eventStr.postfix(regexp, postfixCallback); 
        beforeOn.key('test', keyCallback);
      }
    }
    const Emitter = new EmisorCore({
      plugins: [
        new Plugin
      ]
    });

    Emitter.on('test?<1>', () => {});

    expect(postfixCallback).toBeCalledWith('<1>');
    expect(keyCallback).toBeCalledWith(
      expect.objectContaining({
        options: '1'
      }),
      expect.any(Object)
    );
  });
});

describe.each([
  ['all', []],
  ['key', ['t'], {t:1}],
])('beforeOn hooks', (method, params, options) => {
  test (`beforeOn.${method}`, () => {
    let event = Symbol(),
        handler = () => {},
        callback = jest.fn(),
        checkOptions = options ? options.t : null;

    const Emitter = new EmisorCore({
      plugins: [{
        install({beforeOn}) {
          beforeOn[method](...params, callback);
        }
      }]
    });

    Emitter.on(event, handler, options);

    expect(callback).toBeCalledWith(
      expect.objectContaining({
        event: expect.objectContaining({
          event,
          handler
        }),
        ...checkOptions
      }),
      expect.any(Object)
    );
  });

  test (`beforeOn.${method} break`, () => {
    let  callback = jest.fn(() => ({
          [EmisorPlugin.BREAK_KEY]: true
        })),
        callback2 = jest.fn();

    const Emitter = new EmisorCore({
      plugins: [{
        install({beforeOn}) {
          beforeOn[method](...params, callback);
        }
      }, {
        install({beforeOn}) {
          beforeOn[method](
            ...(params.length ? ['t2'] : params),
            callback2
          );
        }
      }]
    });

    Emitter.on(
      'test', 
      () => {}, 
      options ? { t2: true, ...options} : options
    );

    expect(callback).toBeCalled();
    expect(callback2).not.toBeCalled();

  });
});