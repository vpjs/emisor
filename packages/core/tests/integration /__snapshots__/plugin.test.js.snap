// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin hook interface plugin install hook signature 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "afterOff": Object {
          "all": [Function],
          "key": [Function],
        },
        "afterOn": Object {
          "all": [Function],
          "key": [Function],
        },
        "afterPublish": Object {
          "all": [Function],
          "key": [Function],
        },
        "beforeOff": Object {
          "all": [Function],
          "key": [Function],
        },
        "beforeOn": Object {
          "all": [Function],
          "key": [Function],
        },
        "beforePublish": Object {
          "all": [Function],
          "key": [Function],
        },
        "eventStr": Object {
          "postfix": [Function],
          "prefix": [Function],
        },
        "onEmit": Object {
          "all": [Function],
        },
      },
      Object {
        "emit": [Function],
        "off": [Function],
        "on": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
